File: leetcode-solutions/ruby/2. Add Two Numbers.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
    result = []
    num1, num2 = l1, l2
    list = nil
    remainder = 0
    
    while num1 || num2
        a, b = 0, 0
        
        if num1
            a = num1.val
            num1 = num1.next
        end
        
        if num2
            b = num2.val
            num2 = num2.next
        end
        sum = a + b + remainder
        
        if sum > 9
            sum %= 10
            remainder = 1
        else
            remainder = 0
        end
        
        result << sum
    end
    
    result << remainder if remainder != 0
    
    return nil if result.size < 0
    head = ListNode.new(result.first)
    list = head
    result.each_with_index do |n, i|
        next if i == 0
        
        list.next = ListNode.new(n)
        list = list.next
    end
    
    return head
    
end

File: leetcode-solutions/python/2. Add Two Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        rem = 0
        new_head = cur = ListNode()

        while l1 or l2:
            first_digit = l1.val if l1 else 0
            second_digit = l2.val if l2 else 0
            total = first_digit + second_digit + rem
            cur.next = ListNode(total % 10)
            cur = cur.next
            rem = total // 10
            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next

        # if we have rem
        if rem > 0:
            cur.next = ListNode(rem)

        return new_head.next
